### PR TITLE
[MRG] Wasserstein distance on the circle and Spherical Sliced-Wasserstein

### PR DESCRIPTION
**Pull request by clbonet**
## Types of changes

New functions in solver_1d.py:
- binary_search_circle: binary search to compute the p-Wasserstein distance on the circle (implementation of the algorithm of [Fast transport optimization for Monge costs on the circle](https://arxiv.org/abs/0902.3527)).
- wasserstein1_circle: 1-Wasserstein distance on the circle using the closed-form with the level median as proposed in [The statistics of circular optimal transport](https://arxiv.org/abs/2103.15426).
- wasserstein_circle: Compute the p-Wasserstein distance on the circle using the closed-form for p=1 and the binary search otherwise.
- wasserstein2_unif_circle: Closed-form of the 2-Wasserstein distance between samples and a uniform distribution
- Cost: Compute the p-Wasserstein objective (used in the binary search)
- dCost: Compute the left and right derivative of the p-Wasserstein objective (used in the binary search)
- roll_cols: a util function to roll each row

New function in sliced.py:
- sliced_wasserstein_sphere: Compute the [Spherical Sliced-Wasserstein](https://arxiv.org/abs/2206.08780) discrepancy
- sliced_wasserstein_sphere_unif: SSW_2 between samples and a uniform distribution using the closed-form of the Wasserstein distance

Modifications of backend.py:
- Added some methods in the backend: tile, floor, prod, sort2, qr, atan2, transpose

New function in utils.py:
- get_coordinate_circle: Get the coordinates in turn (in [0,1[) from the points on the circle in the ambient space

New examples:
- plot_compute_wasserstein_circle.py: Example of using ot.wasserstein_circle between von Mises distributions, and ot.wasserstein2_unif_circle
- plot_variance_ssw.py: Example of using ot.sliced_wasserstein_sphere
- plot_ssw_unif_torch.py: Gradient descent on particles to learn a uniform distribution

New tests in test_1d_solver.py:
- test_wasserstein_1d_circle: Compare the values when computing the transport with ot.emd2 and the geodesic distance as ground cost
- test_wasserstein1d_circle_devices: Test devices (except for tensorflow)
- test_wasserstein_1d_unif_circle: Compare the values of ot.wasserstein2_unif_circle and an approximation with ot.emd2 or ot.wasserstein_circle
- test_wasserstein1d_unif_circle_devices: Test devices

New tests in test_sliced.py:
- test_projections_stiefel: Check that the projections are well on the Stiefel manifold
- test_sliced_sphere_same_dist: Check that SSW(x,x)=0
- test_sliced_sphere_bad_shapes: Check the error when the shapes are differents
- test_sliced_sphere: Check the error when the samples are not on the sphere
- test_sliced_sphere_log: Check that log returns the projections and the Wasserstein distance between the projected samples
- test_sliced_sphere_different_dists: Check that SSW is not equal to 0 between different samples
- test_1d_sliced_sphere_equals_emd: Check that we recover the Wasserstein distance on the circle when samples are in S^1
- test_sliced_sphere_backend_type_devices: Test devices (except for tensorflow)
- test_sliced_sphere_unif: Check the error when the samples are not on the sphere
- test_sliced_sphere_unif_log: Check that log returns the projections and the Wasserstein distance between the projected samples
- test_sliced_sphere_unif_backend_type_devices: Test devices

New tests in test_backend.py:
- Added the tests for tile, floor, prod, sort2, qr, atan2 and transpose

New test in test_utils.py:
- test_get_coordinate_circle: Test that we recover the right coordinates

## Motivation and context / Related issue

The Wasserstein distance on the circle can be computed more efficiently using its "closed-forms". For p=1, a closed-form was proposed in [The statistics of circular optimal transport](https://arxiv.org/abs/2103.15426), and for p>=1, a binary search algorithm was proposed in [Fast transport optimization for Monge costs on the circle](https://arxiv.org/abs/0902.3527).

For data lying on the sphere, we can use a particular sliced-Wasserstein discrepancy by using geodesic projections on the circle.

## How has this been tested (if it applies)

- The computation of Wasserstein on the sphere has been compared with ot.emd2 with geodesic ground cost. For the p=1 implementation with the level median, the results seem a bit less precise than with the binary search (the rtol is reduced in the test) but it is quicker.
- The spherical sliced-Wasserstein was tested with the same type of test as sliced-Wasserstein
- I did not test with tensorflow

## PR checklist

- [x] I have read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

**Review by rflamary — COMMENTED** [review]
Hello @clbonet 

This is an impressive PR! I have found a few comments that need to be taken into account (or at leest discussed). Could you look into that please?

File: ot/lp/solver_1d.py [hunk 1]
```diff
@@ -365,3 +365,563 @@ def emd2_1d(x_a, x_b, a=None, b=None, metric='sqeuclidean', p=1., dense=True,
...
+    .. math::
+        W_p^p(u,v) = \inf_{\theta\in\mathbb{R}}\int_0^1 |F_u^{-1}(q)  - (F_v-\theta)^{-1}(q)|^p\ \mathrm{d}q
+
+    where:
+
+    - :math:`F_u` and :math:`F_v` are respectively the cdfs of :math:`u` and :math:`v`
+
+    Parameters
+    ----------
+    u_values : ndarray, shape (n, ...)
```
[the remark this answers is not in the document]
Comment (clbonet) [reply ot/lp/solver_1d.py:594]: In this implementation, we model the circle as $[0,1[$, and hence the function takes real values %1. 

If we want to take as input directly coordinate on $S^1$, then we should add a projection to get the coordinates (as I do in the sliced_wasserstein_sphere for example).

[the remark this answers is not in the document]
Comment (rflamary) [reply ot/lp/solver_1d.py:594]: ok so you should be more clear in teh documentation that u are in [0,1], btw why not [0,2\pi] ?

[the remark this answers is not in the document]
Comment (clbonet) [reply ot/lp/solver_1d.py:594]: Ok, I will add it.

it on [0,1[ because they use this in the reference articles ^^. But, to passe from [0,2\pi[ to [0,1[, we only need to divide by /(2\pi)

File: ot/lp/solver_1d.py [hunk 2]
```diff
@@ -365,3 +365,563 @@ def emd2_1d(x_a, x_b, a=None, b=None, metric='sqeuclidean', p=1., dense=True,
...
+    sum_weights = nx.cumsum(weight_sorted, axis=0) - 0.5
+    sum_weights[sum_weights < 0] = np.inf
+    inds = nx.argmin(sum_weights, axis=0)
+
+    levMed = nx.take_along_axis(cdf_diff_sorted, nx.reshape(inds, (1, -1)), 0)
+
+    return nx.sum(delta * nx.abs(cdf_diff - levMed), axis=0)
+
+
+def wasserstein_circle(u_values, v_values, u_weights=None, v_weights=None, p=1,
```
[the remark this answers is not in the document]
Comment (clbonet) [reply ot/lp/solver_1d.py:806]: Yes, we do not need the two functions. I will juste leave wasserstein_circle public.

File: test/test_1d_solver.py [hunk 3]
```diff
@@ -218,3 +218,108 @@ def test_emd1d_device_tf():
...
+
+        w1 = ot.wasserstein_circle(xb, xb, rho_ub, rho_vb, p=1)
+        w2_bsc = ot.wasserstein_circle(xb, xb, rho_ub, rho_vb, p=2)
+
+        nx.assert_same_dtype_device(xb, w1)
+        nx.assert_same_dtype_device(xb, w2_bsc)
+
+
+def test_wasserstein_1d_unif_circle():
+    # test wasserstein unif_circle give similar results as wasserstein1d
```
[the remark this answers is not in the document]
Comment (clbonet) [reply test/test_1d_solver.py:284]: It is a bit different since I use a lot more samples for v to try to approximate the uniform distribution.

[the remark this answers is not in the document]
Comment (rflamary) [reply test/test_1d_solver.py:280]: ok, i'm just saying you dont' compare to wasserstein1d that is on the line because it is not comparable (see the comment)

File: ot/lp/solver_1d.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (rflamary) [reply ot/lp/solver_1d.py:605]: I know ;). But to get [0,1] from any numpy function (atan or other) you do need to divide by 2pi ;)

I any case i'm OK wit 0,1 the same noation as the paper makes sens just be very clear in the doc because it is highly non standard way to model what is basically a radian

File: test/test_1d_solver.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (clbonet) [reply test/test_1d_solver.py:280]: Ah yes sorry, I misread. I will update the comment.

File: ot/lp/solver_1d.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (clbonet) [reply ot/lp/solver_1d.py:605]: Ok, I tried to be more clear in the doc. Let me know if it is enough clear.
